keps/sig-api-machinery/1904-efficient-watch-resumption/README.md

```diff
@@ -0,0 +1,485 @@
+only be updated to a value that some object of that type had), the quorum
+list, even though it is only requesting objects of that resource type,
+returns "globally current version for the whole etcd".
+
+While the inability to reestablish watch doesn't happen often in practice,
+the kube-apiserver restarts obviously does (e.g. on upgrades). And this
+is resulting in two main problems:
+
+1. In case of rolling upgrade of kube-apiserver, when no object of a given
+   resource type is changing (but objects of other types do), all watchers
+   will eventually be forced to relist, causing significant performance
+   and scalability issues for larger clusters. See the example below for
+   clarification.
+
+1. If no object of a given resource type is changing after kube-apiserver
+  startup, watch cache may stuck being synced to resource version not being
+	version of any object of a given type for extended period of time. See
+  https://github.com/kubernetes/kubernetes/issues/91073 for more details.
+
+To illustrate the problem better, imagine the following example:
+- there is a single kube-apiserver that we are going to restart
+- there are only two resource types: X and Y (for simplicity)
+- there is one object of type X: x with rv=100
+- there is one object of type Y: y with rv=101
+- there is a watch W for objects of type X already synced to rv=100
+- watchcache for both X and Y is up-to-date
+
+- T1: kube-apiserver being restarted
+- T2: kube-apiserver is initialized with watchcache for X synced to
+   rv=101 (most recent etcd version)
+- T3: watch W is trying to reconnect, given rv=100 is outside of
+   supported history window (we only support versions from 101 being
+   the moment of watch cache initialization), relist is forced
+
+Note that adding more kube-apiservers doesn't solve the problem. In
+fact, it is introducing the second problem where watch cache in different
+instances may be out of sync for long period of time. To achieve such
+situation it is enough to have object y be updated to rv=102 between
+the moment kube-apiserver-1 is initialized and kube-apiserver-2 starts
+its initialization. And we would stuck with kube-apiserver-1 synced to
+rv=101 and kube-apiserve-2 synced to rv=102 until the next change of
+any object of type X happens (even though both watch caches contain the
+same state).
+
+From the first glance it may look like having many kube-apiservers can
+mitigate this problem. However, if no object of type X is being changed
+for extended period of time, this doesn't help and I've seen multiple
+large production clusters with tens of thousands of pods for which any
+pod was changing for long enough periods of time (e.g. tens of minutes).
+
+The goal of this proposal is to avoid both of these problems.
+
+### Goals
+
+- avoid tons of relists during kube-apiservers rolling upgrades
+- avoid different instances of kube-apiserver stuck with watchcache synced
+  to different resource versions for extended period of time
```

Comment:
Ah, OK, now I understand what the second problem is. I think the example needs more detail about when the cache "syncs" to the global index and when in "syncs" to the highest index of an object in the collection.

Can you make the problem example more extreme like this? Imagine a collection with objects a, b, c, having RVs 1, 2, 5000 respectively. Now object c is deleted at an inconvenient time while an apiserver is starting. Does this apiserver now believe it is only synced to RV=2 for this collection?

Reply:
Actually it's not the case.

Watchcache claims to be synced to RV of the latest change of the object of a given type. Deletions are handled correctly (so e.g. if the deletion was at RV=6000), it will claim to be synced to 6000, not 2. So that works fine already.

I clairified that in line 89 [I though I wrote that already, but apparently I it wasn't the case.]

The problem is purely with initialization, because initialization is happening NOT from the objects of a given type, but from "global etcd version" (which doesn't have to represent any object of our type).

Reply:
OK, that's good to know. The problem I guess is that deletion records aren't loaded on startup.

Reply:
That isn't really it. Because the RV of the LIST, doesn't depend on what was happening before.
And we don't load events at all (if we would, deletions would load correctly, since these are regular watch events).